sdk/core/System.ClientModel/src/Internal/PipelineMessageSanitizer.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace System.ClientModel.Internal;
+
+/// <summary>
+/// TODO.
+/// </summary>
+internal class PipelineMessageSanitizer
+{
+    private const string LogAllValue = "*";
+    private readonly bool _logAllHeaders;
+    private readonly bool _logFullQueries;
+    private readonly string _redactedPlaceholder;
+    private readonly HashSet<string> _loggedHeaderNames;
+    private readonly IEnumerable<string> _loggedQueryParameters;
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="allowedQueryParameters"></param>
+    /// <param name="allowedHeaders"></param>
+    /// <param name="redactedPlaceholder"></param>
+    public PipelineMessageSanitizer(IEnumerable<string> allowedQueryParameters, IEnumerable<string> allowedHeaders, string redactedPlaceholder = "REDACTED")
+    {
+        _logAllHeaders = allowedHeaders.Contains(LogAllValue);
+        _logFullQueries = allowedQueryParameters.Contains(LogAllValue);
+
+        _loggedQueryParameters = allowedQueryParameters;
+        _redactedPlaceholder = redactedPlaceholder;
+        _loggedHeaderNames = new HashSet<string>(allowedHeaders, StringComparer.InvariantCultureIgnoreCase);
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="name"></param>
+    /// <param name="value"></param>
+    /// <returns></returns>
+    public string SanitizeHeader(string name, string value)
+    {
+        if (_logAllHeaders || _loggedHeaderNames.Contains(name))
+        {
+            return value;
+        }
+
+        return _redactedPlaceholder;
+    }
+
+    /// <summary>
+    /// TODO.
+    /// </summary>
+    /// <param name="url"></param>
+    /// <returns></returns>
+    public string SanitizeUrl(string url)
+    {
+        if (_logFullQueries)
+        {
+            return url;
+        }
+
+#if NET5_0_OR_GREATER
```

Comment:
nit:
```suggestion
#if NET6_0_OR_GREATER
```

In SCM, we don't have a net5.0 target.